contracts/ERC20.sol

```diff
@@ -2,33 +2,208 @@
+
+     /*///////////////////////////////////////////////////////////////
+                        Mappings:
+    //////////////////////////////////////////////////////////////*/
+    mapping(address => uint256) private balances;
+
+    mapping(address => mapping(address => uint)) private allowances;
+
+    mapping(address => uint256) private userMintTracker;
+
+
+         /*///////////////////////////////////////////////////////////////
+                        Events:
+    //////////////////////////////////////////////////////////////*/
+    event AdminTransfer(address oldAdmin, address newAdmin);
+
+     /*///////////////////////////////////////////////////////////////
+                          Modifiers
+    //////////////////////////////////////////////////////////////*/
+
+    modifier onlyAdmin() {
+      require(msg.sender == admin_, "VTZY: Unauthorised user.");
+      _;
+    }
+
+    modifier safeMintAmount(uint256 amount) {
+      require(userMintTracker[msg.sender] + amount <= maxSupplyMintPerWallet, "VTZY: Total minting amount exceeds limit.");
+      require(totalSupply_ + amount <= maxSupply, "VTZY: Total supply exceeds max supply.");
+      _;
+    }
+
+         /*///////////////////////////////////////////////////////////////
+                          Constructor
+    //////////////////////////////////////////////////////////////*/
+    constructor(string memory _name, string memory _symbol) {
+      name_ = _name;
+      symbol_ = _symbol;
+      admin_ = msg.sender;
+      _mint(msg.sender, 1 ether);
+    }
+
+             /*///////////////////////////////////////////////////////////////
+                          Main Functions
+    //////////////////////////////////////////////////////////////*/
 
   function approve(address spender, uint256 amount) external returns (bool) {
+      allowances[msg.sender][spender] = amount;
+
+      emit Approval(msg.sender, spender, amount);
+      return true;
+  }
+
+  function transfer(address to, uint256 amount) external returns (bool) {
+    address addressToTransfer = msg.sender;
+    _transfer(addressToTransfer, to, amount);
     return true;
   }
 
+  
   function transferFrom(
     address from,
     address to,
     uint256 amount
   ) external returns (bool) {
+    address spender = msg.sender;
+    _allowanceExpenditure(from, spender, amount);
+    _transfer(from, to, amount);
+
     return true;
   }
+
+  function mint(uint256 amount) external safeMintAmount(amount){
+    address owner = msg.sender;
+    userMintTracker[owner] += amount;
+    _mint(owner, amount);
+  }
+
+  function adminMint(uint256 amount) external onlyAdmin {
+    address owner = msg.sender;
+    require(totalSupply_ + amount <= maxSupply, "VTZY: Total supply exceeds max supply.");
+    _mint(owner, amount);
+  }
+
+  function burn(address user, uint256 amount) external {
+    if(user == msg.sender) {
+      _burn(msg.sender, amount);
+    } else {
+      _allowanceExpenditure(user, msg.sender, amount);
```

Comment:
Interesting. This part is not wrong, but I have never seen it haha. Probably because when ppl approve for another person, their intention is for these tokens to be spent. If the person being approved wants to burn, he can simply transfer back the token to burn